freeipa/src/main/java/com/sequenceiq/freeipa/flow/freeipa/upscale/handler/ValidateInstancesHealthHandler.java

```diff
@@ -0,0 +1,127 @@
+package com.sequenceiq.freeipa.flow.freeipa.upscale.handler;
+
+import static com.sequenceiq.freeipa.flow.freeipa.upscale.UpscaleFlowEvent.UPSCALE_VALIDATE_NEW_INSTANCES_HEALTH_FINISHED_EVENT;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.google.common.base.Joiner;
+import com.sequenceiq.cloudbreak.common.event.Selectable;
+import com.sequenceiq.cloudbreak.logger.MDCBuilder;
+import com.sequenceiq.flow.event.EventSelectorUtil;
+import com.sequenceiq.flow.reactor.api.handler.ExceptionCatcherEventHandler;
+import com.sequenceiq.flow.reactor.api.handler.HandlerEvent;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.instance.InstanceStatus;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.health.NodeHealthDetails;
+import com.sequenceiq.freeipa.client.FreeIpaClientException;
+import com.sequenceiq.freeipa.entity.InstanceMetaData;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.flow.freeipa.upscale.event.UpscaleFailureEvent;
+import com.sequenceiq.freeipa.flow.freeipa.upscale.event.ValidateInstancesHealthEvent;
+import com.sequenceiq.freeipa.flow.stack.StackEvent;
+import com.sequenceiq.freeipa.service.stack.FreeIpaInstanceHealthDetailsService;
+import com.sequenceiq.freeipa.service.stack.StackService;
+import com.sequenceiq.freeipa.service.stack.instance.InstanceMetaDataService;
+
+import reactor.bus.Event;
+
+@Component
+public class ValidateInstancesHealthHandler extends ExceptionCatcherEventHandler<ValidateInstancesHealthEvent> {
+
+    public static final String PHASE = "Instance health validation";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ValidateInstancesHealthHandler.class);
+
+    @Inject
+    private InstanceMetaDataService instanceMetaDataService;
+
+    @Inject
+    private FreeIpaInstanceHealthDetailsService healthService;
+
+    @Inject
+    private StackService stackService;
+
+    @Override
+    public String selector() {
+        return EventSelectorUtil.selector(ValidateInstancesHealthEvent.class);
+    }
+
+    @Override
+    protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<ValidateInstancesHealthEvent> event) {
+        LOGGER.error("Unexpected exception during the validation of FreeIPA instances' health", e);
+        return new UpscaleFailureEvent(resourceId, PHASE, Set.of(), Map.of(), e);
+    }
+
+    @Override
+    protected Selectable doAccept(HandlerEvent<ValidateInstancesHealthEvent> event) {
+        List<NodeHealthDetails> healthDetails = fetchInstancesHealth(event.getData());
```

Comment:
I would add a log line with all the instance ids. New instances are logged in the HostMetadataSetup. Anyway, it may make sense to log them also here to make support easier, or if for some obscure reason not all instanceIds would make it until here.

Reply:
every event should be logged by default, but I can add the extra log

Reply:
Hm, I forgot that. Thank you.